File: javascript/json-transform/src/functions/TransformerFunctionMatchAll.ts
import TransformerFunction from "./common/TransformerFunction";
import { ArgType } from "./common/ArgType";
import FunctionContext from "./common/FunctionContext";

class TransformerFunctionMatchAll extends TransformerFunction {
  constructor() {
    super({
      argsSets: [
        [
          { name: "pattern", type: ArgType.String },
          { name: "group", type: ArgType.Number, defaultValue: 0 },
        ],
      ],
    });
  }

  override async apply(context: FunctionContext): Promise<any> {
    const str = await context.getString(null);
    if (str == null) {
      return null;
    }
    const patternString = await context.getString("pattern");
    if (patternString == null) {
      return null;
    }
    const matcher = new RegExp(patternString, "g");

    const allMatches: string[] = [];
    const group = await context.getInteger("group");
    for (const match of str.matchAll(matcher)) {
      allMatches.push(match[group ?? 0]);
    }
    return allMatches.length == 0 ? null : allMatches;
  }
}

export default TransformerFunctionMatchAll;

File: java/json-transform/src/main/java/ai/bizone/jsontransform/functions/TransformerFunctionMatchAll.java
package ai.bizone.jsontransform.functions;

import ai.bizone.jsontransform.functions.common.*;
import ai.bizone.jsontransform.functions.common.*;

import java.util.ArrayList;
import java.util.Map;
import java.util.regex.Pattern;

public class TransformerFunctionMatchAll extends TransformerFunction {
    public TransformerFunctionMatchAll() {
        super(FunctionDescription.of(
            Map.of(
                "pattern", ArgumentType.of(ArgType.String).position(0),
                "group", ArgumentType.of(ArgType.Number).position(1).defaultValue(0)
            )
        ));
    }
    @Override
    public Object apply(FunctionContext context) {
        var str = context.getString(null);
        if (str == null) {
            return null;
        }
        var patternString = context.getString("pattern");
        if (patternString == null) {
            return null;
        }
        var matcher = Pattern.compile(patternString).matcher(str);
        var group = context.getInteger("group");
        var allMatches = new ArrayList<>();
        while (matcher.find()) {
            allMatches.add(matcher.group(group));
        }
        return allMatches.size() == 0 ? null : allMatches ;
    }
}
